stackslib/src/chainstate/stacks/transaction.rs

```diff
@@ -138,20 +137,94 @@ fn ClarityVersion_consensus_deserialize<R: Read>(
     }
 }
 
+impl StacksMessageCodec for TenureChangeCause {
+    fn consensus_serialize<W: Write>(&self, fd: &mut W) -> Result<(), codec_error> {
+        let byte = (*self) as u8;
+        write_next(fd, &byte)
+    }
+
+    fn consensus_deserialize<R: Read>(fd: &mut R) -> Result<TenureChangeCause, codec_error> {
+        let byte: u8 = read_next(fd)?;
+        TenureChangeCause::try_from(byte).map_err(|_| {
+            codec_error::DeserializeError(format!("Unrecognized TenureChangeCause byte {byte}"))
+        })
+    }
+}
+
+impl StacksMessageCodec for ThresholdSignature {
+    fn consensus_serialize<W: Write>(&self, fd: &mut W) -> Result<(), codec_error> {
+        let compressed = self.R.compress();
+        let bytes = compressed.as_bytes();
+        fd.write_all(bytes)
+            .map_err(crate::codec::Error::WriteError)?;
+        write_next(fd, &self.z.to_bytes())?;
+        Ok(())
+    }
+
+    fn consensus_deserialize<R: Read>(fd: &mut R) -> Result<Self, codec_error> {
+        use p256k1::point::Compressed;
+        use wsts::{Point, Scalar};
+
+        // Read curve point
+        let mut buf = [0u8; 33];
+        fd.read_exact(&mut buf)
+            .map_err(crate::codec::Error::ReadError)?;
+        let R = Point::try_from(&Compressed::from(buf)).map_err(|_| {
+            crate::codec::Error::DeserializeError("Failed to read curve point".into())
+        })?;
+
+        // Read scalar
+        let mut buf = [0u8; 32];
+        fd.read_exact(&mut buf)
+            .map_err(crate::codec::Error::ReadError)?;
+        let z = Scalar::from(buf);
+
+        Ok(Self { R, z })
+    }
+}
+
+impl TenureChangePayload {
+    pub fn validate(&self) -> Result<(), TenureChangeError> {
```

Comment:
This is going to require access to a _lot_ more context, such as access to the chainstate databases.  In fact, the code for this might instead belong in `NakamotoChainState`.